core/src/main/scala/org/http4s/EntityDecoder.scala

```diff
@@ -99,8 +99,12 @@ object EntityDecoder extends EntityDecoderInstances {
 
   private class OrDec[T](a: EntityDecoder[T], b: EntityDecoder[T]) extends EntityDecoder[T] {
     override def decode(msg: Message): DecodeResult[T] = {
-      if (a.matchesMediaType(msg)) a.decode(msg)
-      else b.decode(msg)
+      msg.headers.get(`Content-Type`).map { contentType =>
+        if (a.matchesMediaType(contentType.mediaType)) a.decode(msg)
+        else b.decode(msg)
+      }.getOrElse {
+        a.decode(msg).orElse(b.decode(msg))
```

Comment:
@bryce-anderson Nice catch. I will write out a test in order to validate that assumption but I think that is correct (even though I am not convinced why that needs to be that way, but we don't have control over `Process` from scalaz-stream). In the event that we cannot try more than one decoder what should we do instead. I believe current behavior is to choose the last one but that seemed like fairly arbitrary behavior to me which is why I tried doing something different.